File: src/app/client/js/authz.js
'use strict';

// IMPORTANT: Make sure this module matches the functionality of authz.py.

angular.module('upmark.authz', [])


.factory('AuthzPolicy', function($parse) {

    function Policy(context, rules, errorFactory, aspect) {
        this.rules = rules != null ? rules : {};
        this.context = context != null ? context : {};
        this.errorFactory = errorFactory ? errorFactory : defaultErrorFactory;
        this.aspect = aspect;
    };
    Policy.prototype.declare = function(decl) {
        var expression = decl.expression;
        if (angular.isObject(expression))
            expression = expression[this.aspect] || 'False';

        var rule = new Rule(
            decl.name, expression, decl.description, decl.failure);
        this.rules[rule.name] = rule;
    };
    Policy.prototype.copy = function() {
        return new Policy(
            angular.extend({}, this.context),
            angular.extend({}, this.rules),
            this.error_factory, this.aspect);
    };
    Policy.prototype.derive = function(context) {
        var policy = this.copy();
        angular.extend(policy.context, context);
        return policy;
    };
    Policy.prototype._check = function(ruleName, context) {
        var rule = this.rules[ruleName];
        if (!rule)
            throw new AuthzConfigError("Unknown rule '" + ruleName + "'");
        try {
            if (rule.check(context))
                return true;
        } catch (e) {
            if (e.name == 'AuthzConfigError')
                throw e;
            throw new AuthzConfigError(
                "Error while evaluating " + ruleName + ": " + e);
        }
        context.$failures.push(rule);
        return false;
    };
    Policy.prototype.permission = function(ruleName) {
        var context = angular.extend({}, this.context);
        var that = this;
        context.$authz = function(ruleName) {
            return that._check(ruleName, context);
        };
        context.len = function(arr) {
            return arr.length;
        };
        context.$failures = [];
        var success;
        try {
            success = this._check(ruleName, context);
        } finally {
            that = null;
        }
        return new Permission(ruleName, success, context.$failures);
    };
    Policy.prototype.check = function() {
        var args = Array.prototype.slice.call(arguments);
        var ruleNames = args.filter(angular.isString);
        var options = args.filter(angular.isObject)[0];
        var match = options ? options.match || 'ANY' : 'ANY';
        var checks = ruleNames.filter(function(ruleName) {
            return this.permission(ruleName).valueOf();
        }, this);
        if (match == 'ANY')
            return checks.length > 0;
        else if (match == 'ALL')
            return checks.length >= ruleNames.length;
        else if (match == 'NONE')
            return checks.length <= 0;
        else
            throw new AuthzConfigError("Unknown match type " + match);
    };
    Policy.prototype.verify = function(ruleName) {
        var permission = this.permission(ruleName)
        if (!permission.valueOf())
            throw this.error_factory(permission.toString());
    };


    function Permission(ruleName, success, failures) {
        this.ruleName = ruleName;
        this.success = success;
        this.failures = failures;
    };
    Permission.prototype.valueOf = function() {
        return this.success;
    };
    Permission.prototype.toString = function() {
        if (!!this)
            return "Success"

        return "Failure: " + this.failures.filter(function(rule) {
            return !!rule.failure;
        }).map(function(rule) {
            return rule.failure;
        }).join('; ');
    };


    function Rule(name, expression, description, failure) {
        this.name = name;
        this.description = description;
        this.expression = expression;
        this.failure = failure;
        expression = this.translateExp(expression);
        expression = this.interpolate(expression);
        // Use Angular's expression parser.
        this.getter = $parse(expression);
    };
    Rule.prototype.check = function(context) {
        return this.getter(context);
    };
    Rule.prototype.translateExp = function(expression) {
        // Convert expression syntax into something usable by the expression
        // parser.
        expression = expression.replace(/(^|\W)not($|\W)/g, '$1!');
        expression = expression.replace(/(^|\W)and($|\W)/g, '$1&&$2');
        expression = expression.replace(/(^|\W)or($|\W)/g, '$1||$2');
        expression = expression.replace(/(^|\W)True($|\W)/g, '$1true$2');
        expression = expression.replace(/(^|\W)False($|\W)/g, '$1false$2');
        return expression;
    };
    Rule.prototype.interpolate = function(expression) {
        expression = expression.replace(/@(\w+)/g, '\$authz("$1")');
        return expression;
    };
    Rule.prototype.toString = function() {
        return this.expression;
    };

    // Exception classes recipe by asselin of StackOverflow
    // https://stackoverflow.com/users/1639641/asselin
    // https://stackoverflow.com/a/27724419/320036
    function AccessDenied(message) {
        this.message = message;
        // Use V8's native method if available, otherwise fallback
        if ("captureStackTrace" in Error)
            Error.captureStackTrace(this, AccessDenied);
        else
            this.stack = (new Error()).stack;
    }
    AccessDenied.prototype = Object.create(Error.prototype);
    AccessDenied.prototype.name = "AccessDenied";
    AccessDenied.prototype.constructor = AccessDenied;
    AccessDenied.prototype.toString = function() {
        return this.message;
    };
    function defaultErrorFactory(message) {
        return new AccessDenied(message);
    };

    function AuthzConfigError(message) {
        this.message = message;
        // Use V8's native method if available, otherwise fallback
        if ("captureStackTrace" in Error)
            Error.captureStackTrace(this, AuthzConfigError);
        else
            this.stack = (new Error()).stack;
    }
    AuthzConfigError.prototype = Object.create(Error.prototype);
    AuthzConfigError.prototype.name = "AuthzConfigError";
    AuthzConfigError.prototype.constructor = AuthzConfigError;
    AuthzConfigError.prototype.toString = function() {
        return this.message;
    };

    return Policy;
})

;

File: src/app/server/authz.py
# IMPORTANT: Make sure this module matches the functionality of authz.py.

import logging
import re

from munch import DefaultMunch

from undefined import undefined


log = logging.getLogger('app.authz')


class Policy:
    def __init__(
            self, context=None, rules=None, error_factory=None, aspect=None):
        self.rules = rules if rules is not None else {}
        self.context = context if context is not None else DefaultMunch(
            undefined)
        self.error_factory = error_factory if error_factory else AccessDenied
        self.aspect = aspect

    def declare(self, decl):
        expression = decl['expression']
        if isinstance(expression, dict):
            # Rule has various aspects. Use matching aspect; deny by default.
            expression = decl['expression'].get(self.aspect, 'False')

        rule = Rule(
            decl['name'], expression,
            decl.get('description'), decl.get('failure'))
        self.rules[rule.name] = rule

    def copy(self):
        return Policy(
            DefaultMunch(undefined, self.context),
            self.rules.copy(),
            self.error_factory, self.aspect)

    def derive(self, context):
        policy = self.copy()
        policy.context.update(context)
        return policy

    def _check(self, rule_name, context):
        rule = self.rules.get(rule_name)
        if not rule:
            raise AuthzConfigError("No such rule %s" % rule_name)
        try:
            if rule.check(context):
                return True
        except AuthzConfigError:
            raise
        except Exception as e:
            raise AuthzConfigError(
                "Error while evaluating %s: %s" % (rule_name, e))
        context['_failures'].append(rule)
        return False

    def permission(self, rule_name):
        context = self.context.copy()
        context['_authz'] = lambda rule_name: self._check(rule_name, context)
        context['len'] = lambda xs: len(xs)
        context['_failures'] = []
        success = self._check(rule_name, context)
        return Permission(rule_name, success, context, context['_failures'])

    def check(self, *rule_names, match='ANY'):
        permissions = [self.permission(rule_name) for rule_name in rule_names]
        if log.isEnabledFor(logging.DEBUG):
            for permission in permissions:
                log.debug("%s: %s", permission.rule_name, permission)
        if match == 'ANY':
            return any(permissions)
        elif match == 'ALL':
            return all(permissions)
        elif match == 'NONE':
            return not any(permissions)
        else:
            raise ValueError("Unknown match type %s" % match)

    def verify(self, rule_name):
        permission = self.permission(rule_name)
        log.debug(
            "%s: %s\n\t%s", permission.rule_name, permission,
            permission.context)
        if not permission:
            raise self.error_factory(str(permission))


class Permission:
    def __init__(self, rule_name, success, context, failures):
        self.rule_name = rule_name
        self.success = success
        self.context = context
        self.failures = failures

    def __bool__(self):
        return self.success

    def __eq__(self, other):
        if hasattr(other, 'failures'):
            return self.failures == other.failures
        return bool(self) == other

    def __str__(self):
        if self:
            return "Granted"

        return "Denied: %s" % '; '.join(
            rule.failure for rule in reversed(self.failures)
            if rule.failure)


class Rule:
    def __init__(self, name, expression, description=None, failure=None):
        self.name = name
        self.description = description
        self.failure = failure
        self.expression = expression
        expression = self.translate_exp(expression)
        expression = self.interpolate(expression)
        self._expression = expression

    def check(self, context):
        # This is not secure - but user-defined expressions should not be used.
        # Only the contents of context may be user-provided.
        # print(self.name, context)
        return eval(self._expression, {'__builtins__': {}}, context)

    def translate_exp(self, expression):
        # Grammar is already the same as Python expressions
        return expression

    def interpolate(self, expression):
        expression = re.sub(r'@(\w+)', r'_authz("\1")', expression)
        return expression

    def __str__(self):
        return self.name

    def __repr__(self):
        return "Rule(%r, %r, %r, %r)" % (
            self.name, self.expression, self.description, self.failure)


class AccessDenied(Exception):
    pass


class AuthzConfigError(Exception):
    pass


policy = Policy()
